### PR TITLE
feat(rulesets): add rules for validation uniqueness of tag names

### LINKED ISSUE #2103: Add rule to check uniqueness of `tags.*.name`
**User story.**
As a user, I want to know if the tags defined in my AsyncAPI document (in any place) have a unique names.

**Is your feature request related to a problem?**
The following AsyncAPI document is invalid and Spectral should throw errors due to non-uniqueness of `tags.*.name`.

```yaml
asyncapi: '2.3.0'
...

tags:
  - name: someTag
    ...
  - name: someTag
    ...
```

**Additional context**
Part of #2100

### PR DESCRIPTION
**Pull request by magicmatatjahu**
Fixes https://github.com/stoplightio/spectral/issues/2103

**Checklist**

- [X] Tests added / updated
- [X] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

**Additional context**

Part of #2100 

Add rule for validation uniqueness of tag names:

**Bad Example**

```yaml
tags:
  - name: "Badger"
  - name: "Badger"
```

**Good Example**

```yaml
tags:
  - name: "Aardvark"
  - name: "Badger"
```

### CODE REVIEW

File: packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  const tagNames = tags.map(item => item.name);
+  return tagNames.reduce((acc, item, idx, arr) => {
+    if (arr.indexOf(item) !== idx && acc.indexOf(item) < 0) {
+      acc.push(item);
+    }
+    return acc;
+  }, [] as string[]);
+}
+
+export default createRulesetFunction<Array<{ name: string }>, null>(
```
Comment (marbemac) [inline packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts]: Interesting - so as far as I can tell, this function is equally applicable / useful in the openapi context (and this same rule, with different `given`, would be valuable in the openapi context). @magicmatatjahu @P0lip what do you think re sharing common functions across default rulesets? I know that we have `@stoplight/spectral-functions`, but the functions in there are very generic, while tags uniqueness is a bit more specific but generic across openapi and asyncapi 🤔.

Comment (P0lip) [reply packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts]: I agree, it's something that could be potentially shared.
We could create a directory under rulesets package for shared utils.

File: packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  {
+    input: null,
+    options: null,
+  },
+  function asyncApi2UniquenessTags(targetVal, _, ctx) {
+    if (!targetVal || targetVal.length === 0) return [];
```
Comment (P0lip) [inline packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts]: one can avoid that if you add a schema for the input

File: packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    if (arr.indexOf(item) !== idx && acc.indexOf(item) < 0) {
+      acc.push(item);
+    }
+    return acc;
+  }, [] as string[]);
+}
+
+export default createRulesetFunction<Array<{ name: string }>, null>(
+  {
+    input: null,
```
Comment (P0lip) [inline packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts]:
```suggestion
    input: { 
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: {
            type: 'string',
          },
        },
      },
    },
```

File: packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  },
+  function asyncApi2UniquenessTags(targetVal, _, ctx) {
+    if (!targetVal || targetVal.length === 0) return [];
+
+    const duplicatedTags = getDuplicateTagNames(targetVal);
+    if (!duplicatedTags || duplicatedTags.length === 0) return [];
+
+    return [
+      {
+        message: `Tags contains duplicate tag names: ${duplicatedTags.join(', ')}.`,
```
Comment (P0lip) [inline packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts]: wouldn't you like to report a duplicate for each particular tag?
Say, you have 10 tags, and only one of them is a dupe. In such a case if you pointed at a particular tag, the error would be slightly more useful.

Comment (magicmatatjahu) [reply packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts]: Yeah, good idea. I changed error's message and path to something like that:

```
          message: `"tags" object contains duplicate tag name "someTag".`,
          path: [...ctx.path, '1', 'name'],
```

What do you think? Is it ok for you, or we should only point to the index of tag, not to the tag's name?

File: packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts (hunk 1 above)
Comment (magicmatatjahu) [reply packages/rulesets/src/asyncapi/functions/asyncApi2UniquenessTags.ts]: I created `shared` folder in the `rulesets` packages with `uniquenessTags` rule. I also create `openapi-tags-uniqueness` rule with this same behaviour as for AsyncAPI. If you would like to introduce that rule in separate PR, please let me know, I will remove that.

File: docs/reference/openapi-rules.md [hunk 5]
```diff
@@ -250,6 +250,28 @@ tags:
   - name: "Badger"
 ```
 
+### asyncapi-tags-uniqueness
+
+OpenAPI object should have non-duplicated tag names (identifiers).
```
Comment (P0lip) [inline docs/reference/openapi-rules.md]:
```suggestion
OpenAPI object must not have duplicated tag names (identifiers).
```

File: docs/reference/asyncapi-rules.md [hunk 6]
```diff
@@ -341,6 +341,28 @@ tags:
 
 **Recommended:** No
 
+### asyncapi-tags-uniqueness
+
+Tags have duplicate names (identifiers).
```
Comment (P0lip) [inline docs/reference/asyncapi-rules.md]:
```suggestion
Tags must not have duplicate names (identifiers).
```

File: packages/rulesets/src/oas/index.ts [hunk 7]
```diff
@@ -212,6 +213,17 @@ const ruleset = {
         },
       },
     },
+    'openapi-tags-uniqueness': {
+      description: 'Each tags must have a unique names.',
```
Comment (P0lip) [inline packages/rulesets/src/oas/index.ts]:
```suggestion
      description: 'Each tag must have a unique name.',
```

File: packages/rulesets/src/asyncapi/functions/asyncApi2DocumentSchema.ts [hunk 8]
```diff
@@ -91,7 +91,7 @@ export default createRulesetFunction<unknown, null>(
     input: null,
     options: null,
   },
-  function oasDocumentSchema(targetVal, _, context) {
+  function asyncApi2DocumentSchema(targetVal, _, context) {
```
Comment (P0lip) [inline packages/rulesets/src/asyncapi/functions/asyncApi2DocumentSchema.ts:97]: good catch 😄

File: packages/rulesets/src/asyncapi/index.ts [hunk 9]
```diff
@@ -358,6 +359,40 @@ export default {
         },
       },
     },
+    'asyncapi-tags-uniqueness': {
+      description: 'Each tags must have a unique names.',
```
Comment (P0lip) [inline packages/rulesets/src/asyncapi/index.ts]:
```suggestion
      description: 'Each tag must have a unique name.',
```

File: packages/rulesets/src/shared/functions/uniquenessTags.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        ...duplicatedTags.map(duplicatedIndex => ({
+          message: `"tags" object contains duplicate tag name "${duplicatedTag}".`,
+          path: [...ctx.path, duplicatedIndex, 'name'],
+        })),
+      );
+    });
+
+    return results;
+  },
+);
```
Comment (P0lip) [inline packages/rulesets/src/shared/functions/uniquenessTags.ts]: How about this?
```suggestion
import { createRulesetFunction, IFunctionResult } from '@stoplight/spectral-core';

type Tags = Array<{ name: string }>;

function getDuplicateTags(tags: Tags): number[] {
  return tags
    .map(item => item.name)
    .reduce<number[]>((acc, item, i, arr) => {
      if (arr.indexOf(item) !== i) {
        acc.push(i);
      }

      return acc;
    }, []);
}

export default createRulesetFunction<Tags, null>(
  {
    input: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: {
            type: 'string',
          },
        },
      },
    },
    options: null,
  },
  function uniquenessTags(targetVal, _, ctx) {
    const duplicatedTags = getDuplicateTags(targetVal);
    if (duplicatedTags.length === 0) return;

    const results: IFunctionResult[] = [];

    for (const duplicatedIndex of duplicatedTags) {
      const duplicatedTag = targetVal[duplicatedIndex].name;

      results.push({
        message: `"tags" object contains duplicate tag name "${duplicatedTag}".`,
        path: [...ctx.path, duplicatedIndex, 'name'],
      });
    }

    return results;
  },
);
```

**Review by P0lip — COMMENTED** [review]
Looking good overall.

File: packages/rulesets/src/shared/functions/uniquenessTags.ts (hunk 10 above)
Comment (magicmatatjahu) [reply packages/rulesets/src/shared/functions/uniquenessTags.ts]: @P0lip Yes, it is much easier, thank you very much! Suggestion applied.

**Comment by magicmatatjahu** [comment]
@P0lip Sorry for such a delay! All suggestions have been applied. Let me know if everything is ok and if it is ok, you can merge when you want :) Unfortunately I had to do a mixed reset + force push (it was the easiest way for me) because the commit linting was showing me errors when I merged from the Github UI `develop` branch to my PR.

File: packages/rulesets/src/shared/functions/uniquenessTags.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  {
+    input: {
+      type: 'array',
+      items: {
+        type: 'object',
+        properties: {
+          name: {
+            type: 'string',
+          },
+        },
```
Comment (P0lip) [inline packages/rulesets/src/shared/functions/uniquenessTags.ts:28]:
```suggestion
        properties: {
          name: {
            type: 'string',
          },
        },
        required: ['name'],
```

**Review by P0lip — DISMISSED** [review]
Looking good 🚀

**Review by P0lip — APPROVED** [review]

**Pull request merged**